syntax = "proto3";

package hipstershop;

option go_package = "./pb";

// ----------------- Cart Service -----------------

service CartService {
    rpc AddItem(AddItemRequest) returns (Empty);
    rpc GetCart(GetCartRequest) returns (Cart);
    rpc EmptyCart(EmptyCartRequest) returns (Empty);
}

message CartItem {
    string product_id = 1;
    int32 quantity = 2;
}

message AddItemRequest {
    string user_id = 1;
    CartItem item = 2;
}

message EmptyCartRequest {
    string user_id = 1;
}

message GetCartRequest {
    string user_id = 1;
}

message Cart {
    string user_id = 1;
    repeated CartItem items = 2;
}

message Empty {}

// ----------------- Recommendation Service -----------------

service RecommendationService {
    rpc ListRecommendations(ListRecommendationsRequest) returns (ListRecommendationsResponse);
}

message ListRecommendationsRequest {
    string user_id = 1;
    repeated string product_ids = 2;
}

message ListRecommendationsResponse {
    repeated string product_ids = 1;
}

// ----------------- Product Catalog Service -----------------

service ProductCatalogService {
    rpc ListProducts(Empty) returns (ListProductsResponse);
    rpc GetProduct(GetProductRequest) returns (Product);
    rpc SearchProducts(SearchProductsRequest) returns (SearchProductsResponse);
}

message Product {
    string id = 1;
    string name = 2;
    string description = 3;
    string picture = 4;
    Money price_usd = 5;
    repeated string categories = 6;
}

message ListProductsResponse {
    repeated Product products = 1;
}

message GetProductRequest {
    string id = 1;
}

message SearchProductsRequest {
    string query = 1;
}

message SearchProductsResponse {
    repeated Product results = 1;
}

// ----------------- Shipping Service -----------------

service ShippingService {
    rpc GetQuote(GetQuoteRequest) returns (GetQuoteResponse);
    rpc ShipOrder(ShipOrderRequest) returns (ShipOrderResponse);
}

message GetQuoteRequest {
    Address address = 1;
    repeated CartItem items = 2;
}

message GetQuoteResponse {
    Money cost_usd = 1;
}

message ShipOrderRequest {
    Address address = 1;
    repeated CartItem items = 2;
}

message ShipOrderResponse {
    string tracking_id = 1;
}

message Address {
    string street_address = 1;
    string city = 2;
    string state = 3;
    string country = 4;
    int32 zip_code = 5;
}

// ----------------- Currency Service -----------------

service CurrencyService {
    rpc GetSupportedCurrencies(Empty) returns (GetSupportedCurrenciesResponse);
    rpc Convert(CurrencyConversionRequest) returns (Money);
}

message Money {
    string currency_code = 1;
    int64 units = 2;
    int32 nanos = 3;
}

message GetSupportedCurrenciesResponse {
    repeated string currency_codes = 1;
}

message CurrencyConversionRequest {
    Money from = 1;
    string to_code = 2;
}

// ----------------- Payment Service -----------------

service PaymentService {
    rpc Charge(ChargeRequest) returns (ChargeResponse);
}

message CreditCardInfo {
    string credit_card_number = 1;
    int32 credit_card_cvv = 2;
    int32 credit_card_expiration_year = 3;
    int32 credit_card_expiration_month = 4;
}

message ChargeRequest {
    Money amount = 1;
    CreditCardInfo credit_card = 2;
}

message ChargeResponse {
    string transaction_id = 1;
}

// ----------------- Email Service -----------------

service EmailService {
    rpc SendOrderConfirmation(SendOrderConfirmationRequest) returns (Empty);
}

message SendOrderConfirmationRequest {
    string email = 1;
    OrderResult order = 2;
}

// ----------------- Checkout Service -----------------

service CheckoutService {
    rpc PlaceOrder(PlaceOrderRequest) returns (PlaceOrderResponse);
}

message PlaceOrderRequest {
    string user_id = 1;
    string user_currency = 2;
    Address address = 3;
    string email = 4;
    CreditCardInfo credit_card = 5;
}

message OrderItem {
    CartItem item = 1;
    Money cost = 2;
}

message OrderResult {
    string order_id = 1;
    string shipping_tracking_id = 2;
    Money shipping_cost = 3;
    Address shipping_address = 4;
    repeated OrderItem items = 5;
}

message PlaceOrderResponse {
    OrderResult order = 1;
}

// ----------------- Ad Service -----------------

service AdService {
    rpc GetAds(AdRequest) returns (AdResponse);
}

message AdRequest {
    repeated string context_keys = 1;
}

message AdResponse {
    repeated Ad ads = 1;
}

message Ad {
    string redirect_url = 1;
    string text = 2;
}
